src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java

```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
             throws Exception {
         if (handler instanceof Dispatcher) {
             Dispatcher dispatcher = (Dispatcher) handler;
-            dispatcher.setCiteCompliant(getInfo().isCiteCompliant());
+            Service service = findService(dispatcher, request, response);
+            if (service != null
+                    && (service.getId().equalsIgnoreCase(getInfo().getId())
+                            || service.getId().equalsIgnoreCase(getInfo().getName()))) {
+                dispatcher.setCiteCompliant(getInfo().isCiteCompliant());
+            }
         }
 
         return true;
     }
 
+    private Service findService(
+            Dispatcher dispatcher, HttpServletRequest request, HttpServletResponse response) {
+        // create a new request instance
+        Request req = new Request();
+
+        // set request / response
+        req.setHttpRequest(request);
+        req.setHttpResponse(response);
+
+        // find the service
+        try {
+            return dispatcher.service(req);
```

Comment:
Good idea, but problematic. The service method at the end calls the dispatcher callbacks, which should not happen:
* they would receive events out of order
* they would receive events multiple times (when the dispatcher actually parses the request)

That said, it should be easy to refactour a method that only computes the service descriptor, make it static, and call it from both the existing Dispatcher.service(...) and from here.

I'm worried about POST requests, as the XML body is read while computing the service... it is properly "pushed back" so that the dispatcher can parse it again? You might want to test a GetFeature in XML format, interactively and/or as part of tests.